PocketTutor-Team/pockettutor-app
# Fix multiple bugs related to the M2 feedback (sanitization, UX...)

# Fix multiple bugs about sanitization, UX, etc.

### PR Description
This PR aims to solve and fix multiple bugs that were detected in the M2 grading and after it. I will describe each bug and the way I solve it in a list so that it is easier to understand.

### Input Sanitization 
**Problem**: The text user inputs were not sanitize allowing user to enter very long text that breaks the UI of our application.

=> _Correction_: I use fixed max size (defined in the ressources file integer.xml) to limit the input size of user. For th first and last name, I fixed the size to 20. For all desciption (so tutor and lesson despcription), I fixed the maximum size to be 500. Concerning lesson title, the maximum size is 50 and for lesson review, the maximum size is 200. With the integer.xml file which defines this constants separately, it will be very easy to change them in the future 

###  Price Average 
_Context_: During the tutor profile creation, when the user needs to indicates his price, we display how much he is far away from the average.

**Problem**: The average used to make the computation was a fixed value and did not depend on the price entered by other tutors in the database.

=> _Correction_: I added a function getAveragePrice() in the ListProfileViewModel, that gets all profile, filter the tutor profiles and compute the average. Note that when no tutor is registered in the app, the "average" is a fixed value of 30.-

### Tutor Profile Creation 
**Problem**: When we create a tutor profile, the profile was saved after the first screen (the one shared with students), but if the user stop the profile creation workflow, it will stored an incomplete profile in the Firebase database.

=> _Correction_: The profile is now saved in Firebase only after the complete creation, so after the screen where the tutor select his schedule. Before it is temporarily saved in the current_user field of the ListProfileViewModel.

### System Top App Bar Color
**Problem**: The system top app bar was in dark purple, which is not very discrete.

=> _Correction_: I changed the top app bar to be the same color as the background color (so white for light mode and very dark grey for dark mode).

### UI of the Time Picker
**Problem**: The time picker dialog was not really user friendly and not coordinated with the date picker dialog.

=> _Correction_: I added a background and changed the disposition of the button "Back" and "Ok".

### Time format
**Problem**: When we had a time of with an hour or minute less that 10, we do not have the correct number of characters (8 AM would be displayed as 8:0 instead of 08:00)

=> _Correction_: I added a formatter to the string that stores the time after the time picker.

### Empty Lesson List in Overview Dropdown
**Problem**: The overview dropdowns could be expanded but if the lesson was empty, it displayed a space which is a bit misleading.

=> _Correction_: I added a Text that displays "No Lessons" when the list of lesson is empty.

### Keyboard Dismissal When Filling Tutor Description
**Problem**: It was not easy to dismiss the keyboard when we fill the tutor description field.

=> _Correction_: I replace the "Enter" icon of the keyboard by a "Done" icon that dismiss the keyboad so that it is easy to dismiss.
 
### Deprecated Icons and Divider
**Problem**: We used some depracted components in particular some icons and the Divider composable.

=> _Correction_: The depracted icons are replaced by non-deprecated ones that looks very similar and the Divider composable have been replaced by a HorizontalDivider.

### Tests
In this PR, I also updated some tests to correspond to the bugs fix I made. Moreover, I added some UI and unit tests to be sure the new features works correctly:
- OverviewTest.kt => add a test to be sure the "No Lessons" text is displayed when we exand an empty section
- ListToDosViewModelTest.kt => add tests to be sure the average is computed correctly when we have no tutor, one tutor and multiple tutor.

### How to test this PR
To test this PR, one could run the test I added, or run the MainActivity and check that all bugs are indeed solved as intented.

### Annex: Time Picker UI
(on the left the old time picker UI and on the right, the updated one)
![image](https://github.com/user-attachments/assets/cffdb8d4-c780-4fe7-9be2-12220dcd37bb)      ![image](https://github.com/user-attachments/assets/32d39a72-1e4a-4603-8a3d-81b17416838d)
